File: 279. Perfect Squares.cpp
class Solution {
public:
    // // recursive: TLE
    // int numSquares(int n) {
    //     if(!n) return 0;
    //     if(n == 1) return 1;
    //     int ret = INT_MAX;
    //     for(int i = sqrt(n); i >= 1; i--){
    //         int num = i * i;
    //         if(n == num) return 1;
    //         ret = min(ret, numSquares(n - num) + 1);
    //     }
    //     return ret;
    // }

    // dp: O(n^1.5) time, O(n)space, 38.48%, ok
    int numSquares(int n) {
        if(!n) return 0;
        vector<int> dp(n + 1, INT_MAX);
        for(int i = 1; i <= n; i++){
            int root = sqrt(i);
            if(root * root == i) dp[i] = 1;
            else{
                for(int j = root; j >= 1; j--){
                    int square = j * j;
                    dp[i] = min(dp[i], dp[i - square] + 1);
                }
            }
        }
        return dp[n];
    }
};

File: 279. Perfect Squares.py
import math
class Solution(object):
  
    # dynamic programming: O(N) time & space
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        if int(math.sqrt(n)) ** 2 == n: return 1
        ret = [1 if int(math.sqrt(i)) ** 2 == i else float('inf') for i in xrange(n+1)]
        for i in xrange(n+1):
            if ret[i] == 1: continue
            for j in xrange(1, int(math.sqrt(i))+1):
                ret[i] = min(ret[i], ret[i-j**2] + 1)
        return ret[n]
        
    # BFS?